File: Sklad/label-parser.js
const SkladLabelParser = (() => {
  function parse(code, rules) {
    const raw = String(code || "").trim().toUpperCase();
    if (!raw) {
      return { raw: "", typeCode: "", model: "", colorCode: "", typeName: "", colorName: "" };
    }

    const separator = rules?.separator || "-";
    const parts = raw.split(separator).map((part) => part.trim()).filter(Boolean);
    const keys = rules?.partLabels || ["type", "model", "color"];
    const map = {};

    keys.forEach((key, index) => {
      map[key] = parts[index] || "";
    });

    const typeCode = map.type || "";
    const model = map.model || "";
    const colorCode = map.color || "";

    return {
      raw,
      typeCode,
      model,
      colorCode,
      typeName: rules?.types?.[typeCode] || typeCode,
      colorName: rules?.colors?.[colorCode] || colorCode,
      title: [rules?.types?.[typeCode] || typeCode, model, rules?.colors?.[colorCode] || colorCode]
        .filter(Boolean)
        .join(" · "),
    };
  }

  return { parse };
})();

File: Sklad/sklad_flutter/lib/label_parser.dart
class LabelInfo {
  const LabelInfo({
    required this.raw,
    required this.typeCode,
    required this.model,
    required this.colorCode,
    required this.title,
  });

  final String raw;
  final String typeCode;
  final String model;
  final String colorCode;
  final String title;
}

class LabelParser {
  static const typeNames = {
    'KOM': 'Комод',
    'TUM': 'Тумба',
    'SHK': 'Шкаф',
    'HLD': 'Холодильник',
    'STL': 'Стол',
  };

  static const colorNames = {
    'DUB': 'Дуб',
    'BEL': 'Белый',
    'CHR': 'Чёрный',
    'WAL': 'Орех',
  };

  static LabelInfo parse(String code) {
    final raw = code.trim().toUpperCase();
    final parts = raw.split('-');
    final typeCode = parts.isNotEmpty ? parts[0] : '';
    final model = parts.length > 1 ? parts[1] : '';
    final colorCode = parts.length > 2 ? parts[2] : '';
    final typeName = typeNames[typeCode] ?? typeCode;
    final colorName = colorNames[colorCode] ?? colorCode;
    final title = [typeName, model, colorName].where((p) => p.isNotEmpty).join(' · ');

    return LabelInfo(
      raw: raw,
      typeCode: typeCode,
      model: model,
      colorCode: colorCode,
      title: title.isEmpty ? raw : title,
    );
  }
}
